### DISCUSSION TITLE
How is this better than the lsd tool?

### DISCUSSION THREAD

**Discussion by StewAlexander-com** [eza-community/eza#679] [Q&A, Howto, Help]
I've used the tool [lsd](https://github.com/lsd-rs/lsd) for awhile, what advantages does this have for it?

**Comment by PThorpe92** [comment]
I think it's really a matter of personal preference. Exa is a few years older than LSD and already had a large user base by the time lsd came around, so I imagine that many use it out of familiarity. Both are awesome tools, I have used both each for quite some time. 

This might not be the most helpful answer but I am not sure if there are any significant features we have that lsd does not at this point. If you are curious, maybe try it out for a while and see what you think. :+1:

**Comment by cafkafk** [comment]
As PThorpe said, I think it varies between people. For me personally, one of the biggest reasons can be summed as this:

```
hyperfine "eza -l --icons" "lsd"
Benchmark 1: eza -l --icons
  Time (mean ± σ):       4.3 ms ±   0.3 ms    [User: 2.6 ms, System: 4.7 ms]
  Range (min … max):     3.9 ms …   5.7 ms    364 runs
 
 
Benchmark 2: lsd
  Time (mean ± σ):     101.4 ms ±   2.5 ms    [User: 4.1 ms, System: 8.6 ms]
  Range (min … max):    97.2 ms … 108.2 ms    29 runs
 
Summary
  eza -l --icons ran
   23.67 ± 1.83 times faster than lsd
   ```
   
However, this is probably better suited to a discussion rather than the issue tracker.

**Reply by PurpleMang** [reply]
I'm wondering, is this benchmark in some gigantic folder? lsd gets about 14ms in my pictures folder. Maybe it got improved since this post idk

**Reply by anorthall** [reply]
I've just done several benchmarks on my Mac, and the performance is similar enough that it can be considered totally irrelevant.

On the first run, eza won:
```console
$ hyperfine "eza -l --icons /etc" "lsd -lg /etc" -w 1000 -m 1000
Benchmark 1: eza -l --icons
  Time (mean ± σ):       5.2 ms ±   0.2 ms    [User: 3.4 ms, System: 14.8 ms]
  Range (min … max):     4.7 ms …   7.3 ms    1000 runs

Benchmark 2: lsd -lg
  Time (mean ± σ):       6.4 ms ±   0.4 ms    [User: 2.0 ms, System: 3.1 ms]
  Range (min … max):     5.9 ms …   9.0 ms    1000 runs

Summary
  eza -l --icons ran
    1.24 ± 0.09 times faster than lsd -lg
```

And on the second, with the same command (but 500 iterations instead of 1,000), `lsd` won:
```console
$ hyperfine "eza -l --icons /etc" "lsd -lg /etc" -w 500 -m 500 -N
Benchmark 1: eza -l --icons /etc
  Time (mean ± σ):       5.5 ms ±   0.2 ms    [User: 3.5 ms, System: 14.5 ms]
  Range (min … max):     5.1 ms …   7.1 ms    557 runs

Benchmark 2: lsd -lg /etc
  Time (mean ± σ):       3.9 ms ±   0.2 ms    [User: 1.6 ms, System: 0.9 ms]
  Range (min … max):     3.6 ms …   5.8 ms    766 runs

Summary
  lsd -lg /etc ran
    1.40 ± 0.09 times faster than eza -l --icons /etc
```

And I compared `--tree` in `/etc`, where `eza` was the 'clear' winner, at a whopping 7ms faster!

```console
$ hyperfine "eza --tree /etc" "lsd --tree /etc" -w 250 -m 250 -i
Benchmark 1: eza --tree /etc
  Time (mean ± σ):       3.2 ms ±   0.6 ms    [User: 2.3 ms, System: 1.6 ms]
  Range (min … max):     2.8 ms …  12.1 ms    457 runs

Benchmark 2: lsd --tree /etc
  Time (mean ± σ):       9.0 ms ±   0.3 ms    [User: 1.7 ms, System: 7.0 ms]
  Range (min … max):     8.6 ms …  10.6 ms    250 runs

Summary
  eza --tree /etc ran
    2.79 ± 0.57 times faster than lsd --tree /etc
```

**Reply by zzantares** [reply]
someone just needs to clone `gh repo clone NixOS/nixpkgs` and do the benchmark on that or on the `/nix/store` of a nix user... 👀

**Reply by boomshroom** [reply]
Well you asked for it:
```
Benchmark 1: eza -l --icons /nix/store
  Time (mean ± σ):     947.7 ms ±  24.9 ms    [User: 709.0 ms, System: 1575.7 ms]
  Range (min … max):   933.0 ms … 1016.1 ms    10 runs

  Warning: Statistical outliers were detected. Consider re-running this benchmark on a quiet system without any interferences from other programs. It might help to use the '--warmup' or '--prepare' options.

Benchmark 2: lsd -l /nix/store
  Time (mean ± σ):      2.165 s ±  0.010 s    [User: 0.779 s, System: 1.378 s]
  Range (min … max):    2.153 s …  2.186 s    10 runs

Summary
  eza -l --icons /nix/store ran
    2.28 ± 0.06 times faster than lsd -l /nix/store
```
Number of files listed: 140866

I am _not_ running `--tree` in this directory.

**Comment by piechologist** [comment]
My pet peeve: eza can sort file names case-sensitively, lsd can't.

**Comment by Dini-25** [comment]
I like to use the tree view to directories only. I can do this in eza but with lsd i cant as it does not allow to specify the depth of the tree when using the directory only option.

**Comment by ghost** [comment]
lsd --tree took 6.28s
eza --tree took 0.29s

it's your choice!

**Reply by jtackaberry** [reply]
Did you run both with a warm cache?  Or did you run lsd first, and then once the cache was warm, eza?

For me, with a warm cache, those commands in my home directory take about the same time.  (7 runs each, throwing away the first two and averaged, `lsd --tree` was 2.198s and `eza --tree` was 2.233s.)

Edit: of course a couple months has passed, so lsd may have improved since then.

**Reply by cafkafk** [reply]
I remember observing non-linear runtime issues here, so as the size of the directory grows, the effect would be more pronounced.

**Comment by nickchomey** [comment]
Im finding lsd to be vastly faster on my system for things like `--tree`

**Comment by SchrodingerZhu** [comment]
`lsd` is faster on zen5:

 ```
~
❯ hyperfine "eza -l --icons /usr/lib" "lsd /usr/lib"
Benchmark 1: eza -l --icons /usr/lib
  Time (mean ± σ):      21.3 ms ±   1.2 ms    [User: 27.3 ms, System: 78.3 ms]
  Range (min … max):    19.6 ms …  25.3 ms    118 runs

Benchmark 2: lsd /usr/lib
  Time (mean ± σ):      18.9 ms ±   1.6 ms    [User: 8.5 ms, System: 10.4 ms]
  Range (min … max):    16.0 ms …  22.2 ms    141 runs

Summary
  lsd /usr/lib ran
    1.13 ± 0.11 times faster than eza -l --icons /usr/lib
```

and 

```
❯ hyperfine "eza -l --icons" "lsd"
Benchmark 1: eza -l --icons
  Time (mean ± σ):       1.9 ms ±   0.2 ms    [User: 1.8 ms, System: 7.2 ms]
  Range (min … max):     1.4 ms …   2.5 ms    953 runs

  Warning: Command took less than 5 ms to complete. Note that the results might be inaccurate because hyperfine can not calibrate the shell startup time much more precise than this limit. You can try to use the `-N`/`--shell=none` option to disable the shell completely.

Benchmark 2: lsd
  Time (mean ± σ):     264.6 µs ± 105.6 µs    [User: 245.8 µs, System: 305.5 µs]
  Range (min … max):    98.0 µs … 939.2 µs    2623 runs

  Warning: Command took less than 5 ms to complete. Note that the results might be inaccurate because hyperfine can not calibrate the shell startup time much more precise than this limit. You can try to use the `-N`/`--shell=none` option to disable the shell completely.

Summary
  lsd ran
    7.14 ± 2.97 times faster than eza -l --icons
```

**Reply by SchrodingerZhu** [reply]
The difference is even larger with `--tree`:

```
❯ hyperfine -N "eza --icons --tree" "lsd --tree"
Benchmark 1: eza --icons --tree
  Time (mean ± σ):      2.210 s ±  0.081 s    [User: 3.364 s, System: 27.151 s]
  Range (min … max):    2.144 s …  2.405 s    10 runs

Benchmark 2: lsd --tree
  Time (mean ± σ):     991.0 ms ±  14.5 ms    [User: 364.5 ms, System: 624.7 ms]
  Range (min … max):   973.0 ms … 1024.5 ms    10 runs

Summary
  lsd --tree ran
    2.23 ± 0.09 times faster than eza --icons --tree
```

This is run on `zfs` btw.

**Reply by tadeasf** [reply]
```bash
---
Benchmarking 'eza_tree' (10 runs after 1 warmup)...
Benchmark 1: eza --icons --tree ~/
  Time (mean ± σ):     13.242 s ±  0.594 s    [User: 20.136 s, System: 21.278 s]
  Range (min … max):   12.799 s … 14.367 s    10 runs
 
  Warning: Statistical outliers were detected. Consider re-running this benchmark on a quiet system without any interferences from other programs. It might help to use the '--warmup' or '--prepare' options.
 
Storing result for 'eza_tree': Median time = 12.9837392169s
---
Benchmarking 'lsd_tree' (10 runs after 1 warmup)...
Benchmark 1: lsd --tree ~/
  Time (mean ± σ):     10.036 s ±  0.163 s    [User: 3.986 s, System: 5.977 s]
  Range (min … max):    9.873 s … 10.316 s    10 runs
 
Storing result for 'lsd_tree': Median time = 9.95973556274s
```

similar results for me. without warmed up cache though lsd is CONSIDERABLY slower (30s+) meanwhile eza without cache warm up is about 18s. so I'd say it's a matter of preference here, if you want marginally better performance with a bit less --tree features - go with lsd. if your files often change though i'd go with eza.

**Comment by benoit-pierre** [comment]
`lsd` show the wrong sizes / units: calculated with power of 2s, but with SI prefixes (MB, KB).

**Reply by gittycat** [reply]
`lsd` is consistent with the native `ls` on both macOS and linux.
eza is not.

**Test:**
First create a dummy file with 2097152 bytes.
That's `2.0M` using base 2 and `2.1M` using base 10 and rounding.
```sh
truncate -s 2097152  foo.bin
```

Now list it
```sh
ls -lh   foo.bin
lsd -lh foo.bin
eza -l  foo.bin
```
You'll get `2.0M` for ls and lsd but eza gives `2.1M`.

The "correct" result to me is the one that is expected from native tools.

Side Note: Outside of shells, macOS uses base 10 and "K", "M" to display sizes (Finder, System Info, etc). 
Apple changed it from base 2 back in 2009 (OS X Snow Leopard) because everyone was confused with division by 1024.
So in effect macOS shows it one way for normal users, and another way for terminal users like us.

**Reply by benoit-pierre** [reply]
The problem is not the different default (base 2 for ls, SI for eza), the problem is the misleading wrong units used by lsd: standard SI units with base 2 calculations.

**Reply by gittycat** [reply]
Yes, GNU and BSD tools like `ls`, `du`, `df` are displaying the unit incorrectly from the IEC/ISO standard.
These tools predate the standard by decades and they've been used for 50 years on millions of systems. That's not going to change.
`lsd` works the same way as these tools. `eza` doesn't.

**Reply by benoit-pierre** [reply]
> Yes, GNU and BSD tools like ls, du, df are displaying the unit incorrectly from the IEC/ISO standard.

No they don't, they use simple one letter prefixes, even with `--si`, that not the problem. The problem is lsd using the standard units:
```bash
# OK, if confusing
▸ /usr/bin/ls -lh /usr/bin/{ls,lsd}
-rwxr-xr-x 1 root root 139K Apr  9 16:08 /usr/bin/ls
-rwxr-xr-x 1 root root 3.4M Aug 20  2024 /usr/bin/lsd
▸ /usr/bin/ls -lh --si /usr/bin/{ls,lsd}
-rwxr-xr-x 1 root root 143k Apr  9 16:08 /usr/bin/ls
-rwxr-xr-x 1 root root 3.6M Aug 20  2024 /usr/bin/lsd
# Wrong!
▸ /usr/bin/lsd -lh /usr/bin/{ls,lsd}
.rwxr-xr-x root root 139 KB Wed Apr  9 16:08:23 2025  /usr/bin/ls
.rwxr-xr-x root root 3.4 MB Tue Aug 20 20:30:54 2024  /usr/bin/lsd
# Would be okay:
▸ /usr/bin/lsd -lh /usr/bin/{ls,lsd}
.rwxr-xr-x root root 139 KiB Wed Apr  9 16:08:23 2025  /usr/bin/ls
.rwxr-xr-x root root 3.4 MiB Tue Aug 20 20:30:54 2024  /usr/bin/lsd
# And eza:
▸ /usr/bin/eza -lh /usr/bin/{ls,lsd}
Permissions Size User Date Modified Name
.rwxr-xr-x  142k root  9 Apr 16:08  /usr/bin/ls
.rwxr-xr-x  3.6M root 20 Aug  2024  /usr/bin/lsd
▸ /usr/bin/eza -lh --binary /usr/bin/{ls,lsd}
Permissions  Size User Date Modified Name
.rwxr-xr-x  139Ki root  9 Apr 16:08  /usr/bin/ls
.rwxr-xr-x  3.4Mi root 20 Aug  2024  /usr/bin/lsd
```

So eza output is consistent with `ls --si`, while lsd not consitent with ls, with or without `--si`. And I would prefer if eza would show the full unit too (`MB` or `MiB`).

**Reply by beedell-roke** [reply]
@benoit-pierre, I agree, and it is worth an issue there. However, `--size bytes` *can* be retroactively parsed to render the correct unit.

**Comment by tony-sol** [comment]
Struggling with choise between eza and lsd right now.
I'd want to choose eza, but lsd has config file and i like it more than just 'alias with flags', so i'm a bit confused
I really want to see config file for eza, even like ripgrep or bat has - just list of flags

**Reply by tony-sol** [reply]
...although that config file lacks of configs

**Comment by jermanuts** [comment]
https://christiano.dev/til/lsd-ls-replacement/ says `lsd` displays icons for file types, making it so much easier to navigate files directories. But you can make `eza` do the same with `--icons=always`.

**Comment by eabase** [comment]
This is just way better documented! And with all screenshots and example themes. ❤️

**Comment by savchenko** [comment]
`--tree` benchmarks on 12th gen i5 laptop:

```
Benchmark 1: eza --tree --icons=never /etc
  Time (mean ± σ):     119.8 ms ±  15.1 ms    [User: 90.4 ms, System: 67.7 ms]
  Range (min … max):    75.5 ms … 153.9 ms    150 runs
 
Benchmark 2: lsd --icon=never --tree /etc
  Time (mean ± σ):      24.3 ms ±   3.7 ms    [User: 10.5 ms, System: 13.6 ms]
  Range (min … max):    20.3 ms …  36.9 ms    150 runs
 
Benchmark 3: tree /etc
  Time (mean ± σ):      13.3 ms ±   2.6 ms    [User: 7.0 ms, System: 6.2 ms]
  Range (min … max):    11.0 ms …  21.7 ms    253 runs
 
Summary
  tree /etc ran
    1.82 ± 0.45 times faster than lsd --icon=never --tree /etc
    8.98 ± 2.08 times faster than eza --tree --icons=never /etc
```

NVMe, EXT4, Debian 13.
